templates/validator/charts.html

```diff
@@ -154,111 +154,149 @@
+                },
+              }
+            }],
+            series: [
+              {
+                name: "Execution Income",
+                data: executionIncomeHistory,
+                showInNavigator: false
+              },
+              {
+                name: "Consensus Income",
+                data: incomeHistory,
+                showInNavigator: true
+              }
+            ],
+            tooltip: {
+              split: false,
+              shared: true,
+              formatter: (tooltip) => {
+                var text = ``
+                var total = 0
 
-                                // add total if hovered point contains rewards for both EL and CL
-                                if (tooltip.chart.hoverPoints.length > 1) {
-                                    text += `<b>Total:</b> ${getIncomeChartValueString(total, currency, ethPrice)}`
-                                }
+                // date and epochs
+                const startEpoch = timeToEpoch(tooltip.chart.hoverPoint.x)
+                const timeForOneDay = 24 * 60* 60 * 1000
+                const endEpoch = timeToEpoch(tooltip.chart.hoverPoint.x + timeForOneDay) - 1
+                const startDate = luxon.DateTime.fromMillis(tooltip.chart.hoverPoints[0].x)
+                const endDate = luxon.DateTime.fromMillis(epochToTime(endEpoch+1))
+                text += `${startDate.toFormat("MMM-dd-yyyy HH:mm:ss")} - ${endDate.toFormat("MMM-dd-yyyy HH:mm:ss")}<br> Epochs ${startEpoch} - ${endEpoch}<br/>`;
 
-                                return text
-                            }
-                        }
-                    });
-                })
+                // income
+                for (var i = 0; i < tooltip.chart.hoverPoints.length; i++) {
+                  const value = tooltip.chart.hoverPoints[i].y;
+                  text += `<span style="color:${tooltip.chart.hoverPoints[i].series.color}">\u25CF</span>  <b>${tooltip.chart.hoverPoints[i].series.name}:</b> ${getIncomeChartValueString(value, currency, ethPrice)}<br/>`
+                  total += value
+                }
 
-        </script>
+                // add total if hovered point contains rewards for both EL and CL
+                if (tooltip.chart.hoverPoints.length > 1) {
+                  text += `<b>Total:</b> ${getIncomeChartValueString(total, currency, ethPrice)}`
+                }
+
+                return text
+              }
+            }
+          });
+
+          var incomeChartLoaded = false;
+          $("#load-income-btn").on("click", () => {
+            if (incomeChartLoaded) {
+              return
+            }
+            incomeChartLoaded = true
+            const minTime = incomeHistory[0].x / 1000
+            const genesisTimestamp = {{$.ChainGenesisTimestamp}}
+            const minDay = Math.floor((minTime - genesisTimestamp) / (24 * 60 * 60)) // floor shouldn't be necessary but just in case
```

Comment:
Is this calculation required in the frontend? Shouldn't passing 0 as `end_day` always work? The backend should be smart enough for the rest. If it isn't, maybe consider enlightening it?

Reply:
It is not required, but since I already had the last 30 days saved in the frontend I didn't want to fetch unnecessary data (later I realized that removing 30 days from 100s does not really improve performance). Having now reworked the endpoint, I don't do this anymore.